### PR TITLE
[Design proposal] Fix EncoderDecoderModel classes to be more like BART and T5

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

TLDR: recently, we've added some new classes called `SpeechEncoderDecoderModel` (#13186) and `VisionEncoderDecoderModel` (#13874). These are very similar to the existing `EncoderDecoderModel` class, allowing to combine any speech/vision encoder with any text decoder. However, there's currently an issue in the sense that we have 2 flavors of decoder models in the library, namely:

* "shifting afterwards": those that shift the logits inside them, when calculating the loss. Examples are `BertLMHeadModel`, `BertGeneration`, `BigbirdForCausalLM`, `GPT2LMHeadModel`, `GPTNeoForCausalLM`, `RobertaForCausalLM`. For these models, the `input_ids` you provide should be equal to the `labels`, and the shifting occurs as follows:
```
loss = None
if labels is not None:
   # Shift so that tokens < n predict n
   shift_logits = lm_logits[..., :-1, :].contiguous()
   shift_labels = labels[..., 1:].contiguous()
   # Flatten the tokens
   loss_fct = CrossEntropyLoss()
   loss = loss_fct(shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1))
```
* "shifting before": those that don't shift the logits, but expect this shifting to already have happened before providing the `input_ids` to the decoder. Examples are `BartForCausalLM`, `PegasusForCausalLM`, and the recently added `Speech2Text2ForCausalLM` and `TrOCRForCausalLM`. Here, the loss calculation is simply:
```
loss = None
if labels is not None:
   loss_fct = CrossEntropyLoss()
   loss = loss_fct(logits.view(-1, self.config.vocab_size), labels.view(-1))
```

## Proposal
In any case, we would like the `EncoderDecoderModel` classes to automatically create the `decoder_input_ids` based on the `labels` provided by the user, similar to how this is done for "vanilla" seq2seq models as BART and T5. However, in order to support both flavors of decoders in the EncoderDecoderModel classes, there are 2 options, either:

* option 1: work with an if-else statement, that creates the `decoder_input_ids` based on the "type" of decoder (type 1 would require to simply set `decoder_input_ids = labels`, whereas type 2 would require to shift the `decoder_input_ids` to be a shifted version of the `labels`). This way, the loss can be computed inside each decoder individually. The disadvantage is that one would need to keep some sort of an exhaustive list of classes that enumerates the model classes of the 2 flavors of decoders.
* option 2: don't compute the loss inside the decoders themselves, but rather in the `EncoderDecoderModel` classes itself. In that way, one can implement the `EncoderDecoderModel` classes similar to the "vanilla" models T5 and BART, which means: (1) first create the `decoder_input_ids` based on the labels provided by the user (i.e. shifting), (2) next forward these through the decoder, and then (3) use the `logits` of the decoder to compute the loss. 

Both would allow to use the `DataCollatorForSeq2Seq`, and in turn the `Seq2SeqTrainer` for the EncoderDecoderModel classes. This enables us to make demo notebooks to fine-tune these model classes using the `Seq2SeqTrainer`.

This draft PR implements option 2.

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
Thanks a lot for putting this together! I agree with the approach taken and the code does look very nice. The slight breaking change is acceptable IMO, especially since it's to make those classes more usable (and they don't seem to be super used at the moment).

**Review by LysandreJik — APPROVED** [review]
Thank you for working on this. I think this second option is the cleaner of the two, but I am a bit uneasy because of the slight breaking change. I think this change is okay but I would really advocate for a warning message mentioning to the user that the training done in this version is slightly different from the training done in previous versions, which could explain a change in the training results if switching from version to version.

**Review by patil-suraj — APPROVED** [review]
Thanks a lot for working on this. I also like Option 2 as well, this actually aligns encoder-decoder models with the seq2seq API

**Review by LysandreJik — APPROVED** [review]
Let's go with this then and have a refactor of the `generate` method in another PR. Feel free to merge whenever you deem it ready.

**Review by patrickvonplaten — APPROVED** [review]
Cool looks good to me! Just left some nits
